pkg/cloud/scope/rosamachinepool.go

```diff
@@ -0,0 +1,183 @@
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package scope
+
+import (
+	"context"
+
+	"github.com/pkg/errors"
+	"k8s.io/klog/v2"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	rosacontrolplanev1 "sigs.k8s.io/cluster-api-provider-aws/v2/controlplane/rosa/api/v1beta2"
+	expinfrav1 "sigs.k8s.io/cluster-api-provider-aws/v2/exp/api/v1beta2"
+	"sigs.k8s.io/cluster-api-provider-aws/v2/pkg/cloud"
+	"sigs.k8s.io/cluster-api-provider-aws/v2/pkg/logger"
+	clusterv1 "sigs.k8s.io/cluster-api/api/v1beta1"
+	expclusterv1 "sigs.k8s.io/cluster-api/exp/api/v1beta1"
+	"sigs.k8s.io/cluster-api/util/conditions"
+	"sigs.k8s.io/cluster-api/util/patch"
+)
+
+// RosaMachinePoolScopeParams defines the input parameters used to create a new Scope.
+type RosaMachinePoolScopeParams struct {
+	Client          client.Client
+	Logger          *logger.Logger
+	Cluster         *clusterv1.Cluster
+	ControlPlane    *rosacontrolplanev1.ROSAControlPlane
+	RosaMachinePool *expinfrav1.ROSAMachinePool
+	MachinePool     *expclusterv1.MachinePool
+	ControllerName  string
+}
+
+// NewRosaMachinePoolScope creates a new Scope from the supplied parameters.
+// This is meant to be called for each reconcile iteration.
+func NewRosaMachinePoolScope(params RosaMachinePoolScopeParams) (*RosaMachinePoolScope, error) {
+	if params.ControlPlane == nil {
+		return nil, errors.New("failed to generate new scope from nil RosaControlPlane")
+	}
+	if params.MachinePool == nil {
+		return nil, errors.New("failed to generate new scope from nil MachinePool")
+	}
+	if params.RosaMachinePool == nil {
+		return nil, errors.New("failed to generate new scope from nil RosaMachinePool")
+	}
+	if params.Logger == nil {
+		log := klog.Background()
+		params.Logger = logger.NewLogger(log)
+	}
+
+	ammpHelper, err := patch.NewHelper(params.RosaMachinePool, params.Client)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to init RosaMachinePool patch helper")
+	}
+	mpHelper, err := patch.NewHelper(params.MachinePool, params.Client)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to init MachinePool patch helper")
+	}
+
+	return &RosaMachinePoolScope{
+		Logger:                     *params.Logger,
+		Client:                     params.Client,
+		patchHelper:                ammpHelper,
+		capiMachinePoolPatchHelper: mpHelper,
+
+		Cluster:         params.Cluster,
+		ControlPlane:    params.ControlPlane,
+		RosaMachinePool: params.RosaMachinePool,
+		MachinePool:     params.MachinePool,
+		controllerName:  params.ControllerName,
+	}, nil
+}
+
+// RosaMachinePoolScope defines the basic context for an actuator to operate upon.
+type RosaMachinePoolScope struct {
+	logger.Logger
+	client.Client
+	patchHelper                *patch.Helper
```

Comment:
What is `*patch.Helper`? Why not just use SSA?

Reply:
patch helper is a util provided by CAPI that takes care of doing some checks and patching status among other things. It uses SSA under the hood